### PR TITLE
feat(deps): Upgrade controller-runtime to 0.16.x

### PR DESCRIPTION
**Pull request by gansheer**
Ref #5211 

**Release Note**
```release-note
feat(deps): Upgrade controller-runtime to 0.16.x
```

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Review by squakez — APPROVED** [review]

File: go.mod [hunk 1]
```diff
@@ -53,7 +53,7 @@ require (
 	knative.dev/eventing v0.39.3
 	knative.dev/pkg v0.0.0-20231023151236-29775d7c9e5c
 	knative.dev/serving v0.39.3
-	sigs.k8s.io/controller-runtime v0.15.2
+	sigs.k8s.io/controller-runtime v0.16.0
```
Comment (squakez) [inline go.mod:56]: I think we can bump to `0.16.5` in one shot. In theory, the biggest problem would be the incompatibility with the minor. The patch should introduce no breaking changes I guess.

Comment (gansheer) [reply go.mod:56]: That's the plan, but for now I make the minimal change to see if the cache works.

File: go.mod [hunk 2]
```diff
@@ -52,8 +52,8 @@ require (
 	k8s.io/utils v0.0.0-20230505201702-9f6742963106
 	knative.dev/eventing v0.40.3
 	knative.dev/pkg v0.0.0-20240116073220-b488e7be5902
-	knative.dev/serving v0.40.1
-	sigs.k8s.io/controller-runtime v0.15.2
+	knative.dev/serving v0.39.3
```
Comment (squakez) [inline go.mod:55]: Why downgrading this?

Comment (gansheer) [reply go.mod:55]: That's probably a mistake, I will upgrade it again.

**Pull request merged**